File: frontend/src/services/agent.ts
/** Agent 执行追踪 SSE 订阅服务
 *
 * 复用 chat.ts 的 fetch+reader SSE 解析模式。
 */

export interface AgentStepEvent {
  execution_id: string
  step_id: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  title: string
  icon: string
  progress: number
  start_time: string | null
  end_time: string | null
  detail: Record<string, unknown> | null
  error: string | null
}

export interface AgentProgressEvent {
  execution_id: string
  step_id: string
  progress: number
}

export interface AgentFinishEvent {
  execution_id: string
  module: string
  status: 'completed' | 'failed'
  steps: AgentStepEvent[]
  started_at: string
  finished_at: string | null
  summary: string
}

export interface AgentCallbacks {
  onStep: (data: AgentStepEvent) => void
  onProgress: (data: AgentProgressEvent) => void
  onFinish: (data: AgentFinishEvent) => void
  onError: (error: string) => void
}

export function subscribeExecution(
  executionId: string,
  callbacks: AgentCallbacks,
): AbortController {
  const controller = new AbortController()
  const token = localStorage.getItem('token')

  fetch(`/api/agent/execution/${executionId}/stream`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        const err = await response.json().catch(() => ({ detail: '订阅失败' }))
        callbacks.onError(err.detail || '订阅失败')
        return
      }

      const reader = response.body?.getReader()
      if (!reader) {
        callbacks.onError('无法读取SSE流')
        return
      }

      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })

        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        let currentEvent = ''
        for (const line of lines) {
          if (line.startsWith('event: ')) {
            currentEvent = line.slice(7).trim()
          } else if (line.startsWith('data: ')) {
            const dataStr = line.slice(6)
            try {
              const data = JSON.parse(dataStr)
              switch (currentEvent) {
                case 'agent_step':
                  callbacks.onStep(data)
                  break
                case 'agent_progress':
                  callbacks.onProgress(data)
                  break
                case 'agent_finish':
                  callbacks.onFinish(data)
                  break
                case 'error':
                  callbacks.onError(data.error || '执行异常')
                  break
              }
            } catch {
              // 跳过解析失败的行
            }
          }
        }
      }
    })
    .catch((err) => {
      if (err.name !== 'AbortError') {
        callbacks.onError(err.message || '连接失败')
      }
    })

  return controller
}

File: backend/app/schemas/agent.py
"""Agent 子系统共享 Pydantic schema"""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


# ── 步骤声明（模块提供给装饰器） ──────────────────────────────

class StepDef(BaseModel):
    """模块声明步骤定义 —— 传入 @track_execution 的参数"""
    id: str = Field(..., description="步骤唯一标识，如 'recognize', 'damage_analysis'")
    title: str = Field(..., description="步骤中文标题，如 'AI识别品类'")
    description: str = Field(default="", description="步骤简述")
    icon: str = Field(default="", description="emoji 或 Unicode 图标")


# ── 运行时步骤状态（SSE 推送） ──────────────────────────────

class AgentStep(BaseModel):
    """单步运行时状态 —— agent_step 事件的 data payload"""
    id: str
    title: str
    description: str = ""
    status: Literal["pending", "running", "completed", "failed"] = "pending"
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    progress: int = Field(default=0, ge=0, le=100)
    icon: str = ""
    detail: Optional[dict] = Field(default=None, description="步骤特定的结构化数据")
    error: Optional[str] = None


# ── 完整执行摘要（agent_finish 事件） ─────────────────────────

class AgentExecution(BaseModel):
    """完整执行摘要 —— agent_finish 事件的 data payload"""
    execution_id: str
    module: str
    status: Literal["running", "completed", "failed"]
    steps: list[AgentStep]
    started_at: datetime
    finished_at: Optional[datetime] = None
    summary: str = ""


# ── SSE 事件 payload 包装 ──────────────────────────────────

class AgentStepEvent(BaseModel):
    """推送给 SSE 客户端的 agent_step 事件"""
    execution_id: str
    step_id: str
    status: Literal["pending", "running", "completed", "failed"]
    title: str
    icon: str = ""
    progress: int = 0
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    detail: Optional[dict] = None
    error: Optional[str] = None


class AgentProgressEvent(BaseModel):
    execution_id: str
    step_id: str
    progress: int = Field(ge=0, le=100)


class AgentFinishEvent(BaseModel):
    execution_id: str
    module: str
    status: Literal["completed", "failed"]
    steps: list[AgentStep]
    started_at: datetime
    finished_at: Optional[datetime] = None
    summary: str = ""
